hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteOperatorCoordinator.java

```diff
@@ -522,7 +522,10 @@ private boolean commitInstant(String instant, long checkpointId) {
         .flatMap(Collection::stream)
         .collect(Collectors.toList());
 
-    if (writeResults.size() == 0) {
+    // Control whether we can commit on empty batch to start a new instant after each checkpoint
+    boolean allowCommitOnEmptyBatch = conf.get(FlinkOptions.WRITE_ALLOW_COMMIT_ON_EMPTY_BATCH);
+
+    if (writeResults.size() == 0 && !allowCommitOnEmptyBatch) {
```

Comment:
When there is no data in current batch but `WRITE_ALLOW_COMMIT_ON_EMPTY_BATCH` is set as true, we will commit instant directly and start a new instant (thus the write functions can get new instant to write and unblock flushing) 

I noticed that in current implementation when there is no data, the write functions will set the `isConfirming` flag as true and wait the coordinator call `sendCommitAckEvents`.  Not sure if there will be some cases that the write function will be blocked because we do not call `sendCommitAckEvents` and just start a new instant instead.

Reply:
@stream2000, if we store the watermark in commit metadata, does it mean that the watermark of this commit is null? BTW, is the empty commit necessary in the timeline? I don't think that the commits in the timeline could be empty and in the design.

Reply:
> I don't think that the commits in the timeline could be empty and beyond the design

We already have a config `hoodie.allow.empty.commit` that defaults to true.  And its documents says that the config is useful in cases where extra metadata needs to be published regardless e.g tracking source offsets when ingesting data

> does it mean that the watermark of this commit is null?

I don't know exactly how you store the watermark, maybe through extra metadata?  In this case, we can still keep the extra metadata in commit files even with an empty commit (maybe we need to modify the code). 

>  this could increase the possibility that there are several commits in one checkpoint.

I don't understand why there could be a possibility that commits multiple instants in one checkpoint, could you show me a simple case? In my understanding, we can make sure that have only one instant during a checkpoint. 

Thanks for your review, hope to have further discussions about this!

Reply:
Overall looks good, just need to note that spark structured streaming commits the empty instant by default, and there is no chance to turn it off, we need to consider whether to keep in line with it.